cli/flox-rust-sdk/src/providers/services.rs

```diff
@@ -1,8 +1,207 @@
+#[derive(Debug)]
+pub struct ConfigPath {
+    pub path: PathBuf,
+    _tempdir: Option<tempfile::TempDir>,
+}
+
+/// Determines the location to write the `process-compose` config file
+pub fn process_compose_config_write_location() -> Result<ConfigPath, ServiceError> {
+    if let Ok(path) = env::var(SERVICES_TEMP_CONFIG_PATH_VAR) {
+        return Ok(ConfigPath {
+            path: PathBuf::from(path),
+            _tempdir: None,
+        });
+    }
+    let temp_dir = tempfile::tempdir().map_err(ServiceError::WriteConfig)?;
+    let path = temp_dir.path().join("process-compose.yaml");
+    Ok(ConfigPath {
+        path,
+        _tempdir: Some(temp_dir),
+    })
+}
+
+pub fn maybe_make_service_config_file(
+    flox: &Flox,
+    lockfile: &LockedManifestCatalog,
+) -> Result<Option<ConfigPath>, ServiceError> {
+    let service_config_path = if flox.features.services {
+        let config_path = process_compose_config_write_location()?;
+        write_process_compose_config(
+            &lockfile.manifest.services.clone().into(),
+            &config_path.path,
+        )?;
+        tracing::debug!(
+            path = traceable_path(&config_path.path),
+            "wrote service config"
+        );
+        Some(config_path)
+    } else {
+        None
+    };
+    Ok(service_config_path)
+}
+
+#[cfg(test)]
+mod tests {
+    use std::str::FromStr;
+
+    use indoc::indoc;
+    use proptest::prelude::*;
+
+    use super::*;
+    use crate::flox::test_helpers::flox_instance_with_optional_floxhub_and_client;
+    use crate::flox::{EnvironmentName, EnvironmentOwner};
+    use crate::models::environment::path_environment::{InitCustomization, PathEnvironment};
+    use crate::models::environment::{Environment, PathPointer};
+
+    proptest! {
+        #[test]
+        fn test_process_compose_config_roundtrip(config: ProcessComposeConfig) {
+            let path = process_compose_config_write_location().unwrap();
+            write_process_compose_config(&config, &path.path).unwrap();
+            let contents = std::fs::read_to_string(&path.path).unwrap();
+            let deserialized: ProcessComposeConfig = serde_yml::from_str(&contents).unwrap();
+            prop_assert_eq!(config, deserialized);
+        }
+    }
+
+    #[test]
+    fn built_environments_generate_service_config() {
+        let (mut flox, workspace_dir) = flox_instance_with_optional_floxhub_and_client(
+            Some(&EnvironmentOwner::from_str("owner").unwrap()),
+            true,
+        );
+        flox.features.services = true;
+        let temp_dir = tempfile::tempdir().unwrap();
+        let pointer = PathPointer::new(EnvironmentName::from_str("services_env").unwrap());
+        let customization = InitCustomization::default();
+        let mut env = PathEnvironment::init(
+            pointer,
+            workspace_dir.path(),
+            temp_dir.path(),
+            &flox.system,
+            &customization,
+            &flox,
+        )
+        .unwrap();
```

Comment:
nonblocking: use `new_path_environment` unless there's a reason not to

Reply:
Done in 8579324.